### PR TITLE
Rpc 112 make fill

### PR DESCRIPTION
**Pull request by disenotov**

### CODE REVIEW

File: exchange-v2/contracts/LibOrderDataV2.sol [hunk 1]
```diff
@@ -0,0 +1,22 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.7.6;
+pragma abicoder v2;
+
+import "@rarible/royalties/contracts/LibPart.sol";
+
+library LibOrderDataV2 {
+    bytes4 constant public V2 = bytes4(keccak256("V2"));
+    bytes4 constant public V1 = bytes4(keccak256("V1"));
```
Comment (evgenynacu) [inline exchange-v2/contracts/LibOrderDataV2.sol:10]: is V1 needed here?

File: exchange-v2/contracts/LibOrder.sol [hunk 2]
```diff
@@ -24,9 +24,14 @@ library LibOrder {
...
-    function calculateRemaining(Order memory order, uint fill) internal pure returns (uint makeValue, uint takeValue) {
-        takeValue = order.takeAsset.value.sub(fill);
-        makeValue = LibMath.safeGetPartialAmountFloor(order.makeAsset.value, order.takeAsset.value, takeValue);
+    function calculateRemaining(Order memory order, uint fill, bool isMakeFill) internal pure returns (uint makeValue, uint takeValue) {
+        if (!isMakeFill){
+            takeValue = order.takeAsset.value.sub(fill);
+            makeValue = LibMath.safeGetPartialAmountFloor(order.makeAsset.value, order.takeAsset.value, takeValue);
+        } else {
+            makeValue = order.makeAsset.value.sub(fill);
+            takeValue = LibMath.safeGetPartialAmountFloor(order.takeAsset.value, order.makeAsset.value, makeValue);
```
Comment (evgenynacu) [inline exchange-v2/contracts/LibOrder.sol:33]: probably, we need to use safeGetPartialAmountCeil. need to think about it

File: exchange-v2/test/v2/ExchangeV2.rarible.test.js [hunk 3]
```diff
@@ -18,122 +18,122 @@ const { Order, Asset, sign } = require("../order");
...
-	let transferManagerTest;
-	let t1;
-	let t2;
-	let libOrder;
-	let protocol = accounts[9];
-	let community = accounts[8];
-	const eth = "0x0000000000000000000000000000000000000000";
-	let erc721TokenId0 = 52;
+  let testing;
+  let transferProxy;
```
Comment (evgenynacu) [inline exchange-v2/test/v2/ExchangeV2.rarible.test.js:25]: too many changes in this file. could you revert it and change only needed content? there can be many conflicts in future

**Review by evgenynacu — CHANGES_REQUESTED** [review]

File: exchange-v2/contracts/LibOrderDataV2.sol (hunk 1 above)
Comment (disenotov) [reply exchange-v2/contracts/LibOrderDataV2.sol]: fixed

File: exchange-v2/test/v2/ExchangeV2.rarible.test.js (hunk 3 above)
Comment (disenotov) [reply exchange-v2/test/v2/ExchangeV2.rarible.test.js]: fixed

File: exchange-v2/contracts/LibOrder.sol (hunk 2 above)
Comment (disenotov) [reply exchange-v2/contracts/LibOrder.sol:35]: decided not to fix

**Comment by hantuzun** [comment]
Hi @disenotov & @evgenynacu,

These orders work if and only if the left order on `matchOrders` has `true` for `isMakeFill`.

Is there a reason for this `isMakeFill` functionality that allows Order fills to record either the make asset fills or take asset fills?

It allows recording the fills for the same asset type while matching orders. However, this doesn't have any benefit. What am I missing?

**Comment by hantuzun** [comment]
I've written some tests demonstrating the `matchOrders` failures when the left asset has `false` for `isMakeFills`: https://gist.github.com/hantuzun/5a80e1bada3b8afb75b7b1dc0b339f4a

**Comment by disenotov** [comment]
hi @hantuzun !
i'm not sure if there is a problem

test for the case when both orders have isMakeFill = false is here https://github.com/rarible/protocol-contracts/blob/master/exchange-v2/test/v2/ExchangeV2.rarible.test.js#L969 and it works

i tried changing orders position in this test https://github.com/rarible/protocol-contracts/blob/master/exchange-v2/test/v2/ExchangeV2.rarible.test.js#L929 so it would be for example 
```
testing.matchOrders( right, "0x", left, await getSignature(left, seller), { from: buyer, value: 600, gasPrice: 0 })
```
and it still worked correctly
am i missing something? maybe there is an error in your tests?

**Comment by hantuzun** [comment]
Hi @disenotov!

My tests are based on the order data v2 tests you refer to.

For all 4 possibilities of isMakeTrueThe ERC1155 and ETH amount:
* 2 tests fail and 2 tests succeed when I use the amounts from:
  * https://github.com/rarible/protocol-contracts/blob/master/exchange-v2/test/v2/ExchangeV2.rarible.test.js#L929
* All 4 tests succeed when I use the amounts from:
  * https://github.com/rarible/protocol-contracts/blob/master/exchange-v2/test/v2/ExchangeV2.rarible.test.js#L969

The tests fail with the following stack trace:
```
Reverted with reason string 'rounding error'
    at Exchange.safeGetPartialAmountFloor (LibMath.sol)
    at Exchange.calculateRemaining (LibOrder.sol)
    at Exchange.fillOrder (LibFill.sol)
    at Exchange.getFillSetNew (ExchangeCore.sol)
    at Exchange.matchAndTransfer (ExchangeCore.sol)
    at Exchange.matchOrders (ExchangeCore.sol)
```

In both of the failed cases, the same parameters cause the failure:

`calculateRemaining(order, fill, isMakeFill)` is called where:
 * `order.makeAsset.value` is `200`.
 * `order.takeAsset.value` is `600`.
 * `fill` is `500`.
 * `isMakeFill` is `false`.

Then:
`takeValue = order.takeAsset.value.sub(fill);` sets `takeValue` to `100`.

Finally, the following line reverts:
`makeValue = LibMath.safeGetPartialAmountFloor(order.makeAsset.value, order.takeAsset.value, takeValue);` 

The failing call with replaced values is:
`safeGetPartialAmountFloor(200, 600, 100);`

I'm surprised by the fact that changing the amounts fixes or breaks this test case. I'll share more details soon.

**Comment by hantuzun** [comment]
In the test, the first `matchOrders` happens at a price of `5`.

Then, the remaining sell order updates its price, while keeping the same hash, therefore preserving its fill record.
And then another `matchOrders` is called at the new price for the remaining sell order.

Below is the behavior for the second matchOrder's price:
1: SafeMath: subtraction overflow.
2: SafeMath: subtraction overflow.
3: The rounding error I've noticed the first. As in https://github.com/rarible/protocol-contracts/blob/master/exchange-v2/test/v2/ExchangeV2.rarible.test.js#L929
4: The amount change is 300 instead of 400.
5: OK. The same price as the first match.
6: The rounding error.
8: The rounding error.
9: The rounding error.
10: OK. As in https://github.com/rarible/protocol-contracts/blob/master/exchange-v2/test/v2/ExchangeV2.rarible.test.js#L969
11: The rounding error.
14: The rounding error.
15: The rounding error.
20: OK.
21: The rounding error.
25: OK.
30: The rounding error.
35: The rounding error.
45: The rounding error.
50: OK.
100: OK.
150: The rounding error.
900: The rounding error.
1000: The rounding error.
5000: The rounding error.
10000: The rounding error.

The tests fail more often than not.

You can see how I perform these tests by changing a single variable here at the latest commit: https://gist.github.com/hantuzun/5a80e1bada3b8afb75b7b1dc0b339f4a/revisions?diff=unified

**Comment by hantuzun** [comment]
Changing order positions like the below doesn't change any behavior:
```
testing.matchOrders( right, "0x", left, await getSignature(left, seller), { from: buyer, value: 600, gasPrice: 0 })
```

**Comment by hantuzun** [comment]
What is the purpose of `isMakeFill`?

Also, it'd have helped if running the project tests were documented better.

**Comment by hantuzun** [comment]
The error is just that, a rounding error. 1/6 of 200 NFTs is 33 with integer arithmetics and it is divergent from the actual value of 33.33 by more than 0.1%.

I think `isMakeFill` parameter should be for setting the fill to NFT sides, right?

**Comment by disenotov** [comment]
hi!
yes, the error in your tests is due to the rounding error.

as for the isMakeFill, i started writing an answer and forgot to finish:)
- What is the purpose of isMakeFill?
so, we store orders' `fill` (what part of order is completed) onchain. By default fills are calculated by the take side of the order (e.g. if there is an order 3ERC1155 => 300 ETH, here ERC1155 is the make side, ETH is the take side. it can be matched with another order 100ETH => 1ERC1155, and the first order would be partially filled. so we would store fill = 100 ETH (using take side) for this order , and it can be matched again with 2ERC1155 => 200 ETH for example) 
so isMakeFill = true means that fills are calculated using make-side of the order, in the previous example we would store 1 (for the amount of erc1155 traded) instead of 100 (amount of ETH)
it's usefull for sell orders, because it's important for sellers to sell the right amount of nfts. So the order can be considered filled when all nfts are sold.

about documentation, what do you think is missing?

**Comment by hantuzun** [comment]
Thanks for the clarification!

As I've guessed, `isMakeFill` allows the flexibility for setting the fill side to the NFT asset side. This flexibility can cause rounding errors if the fill side is set to the ETH or ERC20 side by mistake. From my understanding, the fill side should always be the NFT side, and in the case of an order of 2 ERC1155 assets, the fill side should be the ERC1155 with less amount to avert rounding errors.

The purpose and proper usage of `isMakeFill` could be documented.

**Pull request merged**